File: Chapter02/Exercise2_06/Exercise2_06.py
from pyspark.sql import SparkSession
from Chapter02.utilities02_py.helper_python import sample_warc_loc, extract_raw_records, parse_raw_warc
from timeit import default_timer


def heavy_computation():
    list = [1] * 1000
    total_sum = 0
    for i in range(0, 1000):
        total_sum += sum(list)
    return total_sum


if __name__ == "__main__":
    session: SparkSession = SparkSession.builder \
        .appName('Different Tasks') \
        .getOrCreate()
    session.sparkContext.setLogLevel('ERROR')  # avoids printing of info messages

    raw_records = extract_raw_records(sample_warc_loc, session)
    warc_records = raw_records.flatMap(lambda record: parse_raw_warc(record))
    invoked_heavy_rdd = warc_records.map(lambda record: heavy_computation())

    start_time = default_timer()
    invoked_heavy_rdd.foreach(lambda record: None)
    end_time = default_timer()
    print(str(end_time - start_time))

File: Chapter02/Exercise2_06/Exercise2_06.scala
package Exercise2_06

import org.apache.hadoop.io.Text
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.SparkSession
import Utilities02.WarcRecord
import Utilities02.HelperScala.{sampleWarcLoc, extractRawRecords, parseRawWarc}

object Exercise2_06 {

  def heavyComputation(record: WarcRecord): Long = {
    val array = Array.fill(1000)(1)
    var totalSum = 0L
    for (_ <- 0 to 10000) {
      totalSum += array.sum
    }
    totalSum
  }

  def main(args: Array[String]): Unit = {
    implicit val session: SparkSession = SparkSession.builder
      .appName("Different Tasks")
      .getOrCreate()
    session.sparkContext.setLogLevel("ERROR")

    val rawRecords: RDD[Text] = extractRawRecords(sampleWarcLoc)
    val warcRecords: RDD[WarcRecord] = rawRecords.flatMap(parseRawWarc)
    val invokeHeavyRDD = warcRecords.map(heavyComputation)

    session.time(invokeHeavyRDD.foreach(_ => Unit))

  }
}